Complete the porting of icub-main to yarp4

valegagge opened the pull request:
In this PR, we'd like to merge the necessary changes for aligning icub-main to yarp4.

>[!NOTE]
> This code hasn't been tested!

Now we are going to run the CI, then we'll test it on the setup/robot. In this particular case, we do the opposite of what we usually do — we run the CI first and then we run the tests, because the changes are due to yarp modification changes, and any feature has been changed.

See here for more detail. https://github.com/robotology/icub-main/issues/1056

This PR is strictly related to this yarp PR: https://github.com/robotology/yarp/pull/3305

Thanks to @AntonioViscomi, who was the main contributor of these changes!

valegagge commented on src/libraries/icubmod/dragonfly2/linux/FirewireCameraDC1394-DR2_2.cpp:
```diff
@@ -1942,63 +1953,73 @@ bool CFWCamera_DR2_2::setOperationModeDC1394(bool b1394b)
+
+    return bTxON ? yarp::dev::ReturnValue_ok : yarp::dev::ReturnValue::return_code::return_value_error_generic;
+
 }
```
resolved in https://github.com/robotology/icub-main/pull/1058/commits/9039f3ac4ef0d0fa4aa390689c85b8b653657c81

valegagge commented:
/gemini review

valegagge commented on src/libraries/icubmod/dragonfly2/common/DragonflyDeviceDriver2.cpp:
```diff
@@ -254,85 +254,87 @@ bool DragonflyDeviceDriver2::setOnePushDC1394(int feature)
 {
-	return RES(system_resources)->getOperationModeDC1394();
+	b1394 = RES(system_resources)->getOperationModeDC1394(b1394);
+	return ReturnValue_ok;
```
fix in https://github.com/robotology/icub-main/pull/1058/commits/ac0309e7309deca443d8d66e2c8edfd10a563436

AntonioViscomi commented on src/libraries/icubmod/dragonfly2/common/DragonflyDeviceDriver2.h:
```diff
@@ -580,72 +580,72 @@ class yarp::dev::DragonflyDeviceDriver2 :
+    *	unsigned int mask;
     *
-    *	for (unsigned int m=0; m<codings.num; ++m)
+    *	for (unsigned int m; m<codings.num; ++m)
```
done https://github.com/AntonioViscomi/icub-main/commit/21a890609d01be8b3a8e37e7d9a4c52ec9b30323

valegagge commented on src/libraries/icubmod/dragonfly2/linux/FirewireCameraDC1394-DR2_2.cpp:
```diff
@@ -2167,39 +2185,39 @@ bool CFWCamera_DR2_2::setBytesPerPacketDC1394(unsigned int newBandPercent)
     int fps=maxFPS(videoMode,colorCoding);
-    double bpp=bytesPerPixel(colorCoding);
-    double newBandOcc=double(fps*xDim*yDim)*bpp;
+    double bytesPerPixelValue=bytesPerPixel(colorCoding); // And here used for computation. Should we rename this to e.g. "double bpPixel?
```
Fixed in https://github.com/robotology/icub-main/pull/1058/commits/21a890609d01be8b3a8e37e7d9a4c52ec9b30323

valegagge commented on src/libraries/icubmod/dragonfly2/common/DragonflyDeviceDriver2.h:
```diff
@@ -826,48 +826,48 @@ class yarp::dev::DragonflyDeviceDriver2 :
-    FeatureMode TRANSL_MODE(bool mode) { return (mode ? MODE_AUTO : MODE_MANUAL); }
-    bool TRANSL_MODE(FeatureMode mode) { return (mode == MODE_AUTO? 1 : 0); }
+    FeatureMode TRANSL_MODE(bool mode) { return (mode ? FeatureMode::MODE_AUTO : FeatureMode::MODE_MANUAL); }
+    bool TRANSL_MODE(FeatureMode mode) { return (mode == FeatureMode::MODE_AUTO? 1 : 0); }
```
Fixed in https://github.com/robotology/icub-main/pull/1058/commits/21a890609d01be8b3a8e37e7d9a4c52ec9b30323

valegagge commented:
/gemini review

valegagge commented:
Hi @AntonioViscomi !
Hey! We finally did it! We’ve completed the porting!!! Let's go with the test on a setup!

valegagge commented:
Some days ago, I started the tests.
1. the setup works as expected.
2. I created a docker image with all sw on yarp4 and the correposnding fw, modifying also the confing file with the new params. I tested the calibration 12 and position control and they works fine except for these warnings/issues:

<details>
  <summary> YRI crashes due to  `yarp_telemetryDeviceDumper.so`</summary>


  ``` [DEBUG] |yarp.dev.PolyDriver|telemetryDeviceDumper| Parameters are (PC104 (PC104IpAddress "10.0.1.104") (PC104IpPort 12345) (PC104TXrate 1) (PC104RXrate 5)) (auto_save true) (axesNames (torso_pitch_mj1_real_encoder torso_pitch_mj1_fake_encoder)) (device telemetryDeviceDumper) (experimentName torso_pitch_mj1_amo_aksim_temp) (id telemetryDeviceDumper) (logIAmplifierControl true) (logIControlMode false) (logIEncoders true) (logIInteractionMode false) (logIMotorEncoders true) (logIMotorTemperatures true) (logIPidControl false) (logIRawValuesPublisher false) (logITorqueControl false) (log_period 0.00200000000000000004163) (n_samples 80000000) (path "./torso_pitch_mj1_telemetry_data/") (rawValuesPublisherRemoteName "/torso-pitch-mj1-setup/raw_data") (robotName "/torsoPitchMj1") (saveBufferManagerConfiguration true) (save_period 10.0) (save_periodically false) (yarp_robot_name torsoPitchMj1)
[ERROR] yarprobotinterface intercepted a segmentation fault caused by a faulty plugin:
[ERROR] /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os7Vocab326encodeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x13) [0x71308eba5a33]
Trace requested at /home/icub/projects/robotology-superbuild/src/YARP/src/commands/yarprobotinterface/Module.cpp:82 by code called from:
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_Z16yarp_print_traceP8_IO_FILEPKcj+0x3b) [0x71308eb3a6fb]
yarprobotinterface(+0xb114) [0x6037a977f114]
/lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x71308e666520]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os7Vocab326encodeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x13) [0x71308eba5a33]
/home/icub/projects/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN9robometry21TelemetryDeviceDumper22loadSettingsFromConfigERN4yarp2os10SearchableE+0x1c0a) [0x71308427f34a]
/home/icub/projects/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN9robometry21TelemetryDeviceDumper4openERN4yarp2os10SearchableE+0x50) [0x713084280d90]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_dev.so.3(_ZN4yarp3dev10PolyDriver8coreOpenERNS_2os10SearchableE+0x691) [0x71308e5b54f1]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_dev.so.3(_ZN4yarp3dev10PolyDriver4openERNS_2os10SearchableE+0x50) [0x71308e5b6120]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface6Device4openEv+0x5c) [0x71308ecaa5ec]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface5Robot7Private11openDevicesEv+0x4c3) [0x71308ecb2af3]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface5Robot10enterPhaseENS0_11ActionPhaseE+0x637) [0x71308ecb4727]
yarprobotinterface(+0xc127) [0x6037a9780127]
/home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os8RFModule9runModuleERNS0_14ResourceFinderE+0xf7) [0x71308eb8f8d7]
  ```

</details> 

<details>
  <summary> wrong error in `calibrationDoneRaw`: in the end the joint is calibrated </summary>

  ```
DEBUG] |yarp.yri.Calibrator| torso-pitch-mj1-setup-calibrator starting calibration of device torso-pitch-mj1-setup-mc_remapper

[INFO] torsoPitchMj1 : starting calibration

[DEBUG] torsoPitchMj1 : Joints calibration order: 0

[WARNING] torsoPitchMj1  is calibrating only a subset of the robot part. Calibrating  1  over a total of  2

[DEBUG] torsoPitchMj1 : Calling calibrateJoint on joint  0 : type  12  with params:  9162 0 0 0 0

[DEBUG] from BOARD 10.0.1.1 (torso-pitch-mj1-setup) time=467s 555m 596u :  src LOCAL, adr 0,(code 0x04000001, par16 0x0000 par64 0x0000000000000000) -> DEBUG: tag01  CALIB 12 j 0: offset=41930 zero=-32768 

[DEBUG] torsoPitchMj1 : set 1 j 0 : Calibrating... enc values just before starting low-level calibration:  -2.80701

[ERROR] BOARD torso-pitch-mj1-setup (IP 10.0.1.1)  Unable to complete calibration: joint 0  is in  CALIBRATION  status in  virtual yarp::dev::ReturnValue yarp::dev::embObjMotionControl::calibrationDoneRaw(int)

[DEBUG] from BOARD 10.0.1.1 (torso-pitch-mj1-setup) time=468s 139m 65u :  src LOCAL, adr 0,(code 0x04000001, par16 0x0001 par64 0x0000000000000001) -> DEBUG: tag01  SET CONTROLMODE

[DEBUG] torsoPitchMj1 : calib joint  0 ended

[DEBUG] torsoPitchMj1 : set 1 : Calibration ended, going to zero!

[DEBUG] torsoPitchMj1 : Sending positionMove to joint 0  (desired pos:  0 desired speed:  30  )

[DEBUG] torsoPitchMj1: checkGoneToZeroThreshold: joint: 0 curr: -0.522 des: 0.000 -> delta: 0.522 threshold: 2.000 output: 319.000 mode: pos

[DEBUG] torsoPitchMj1: checkGoneToZeroThreshold: joint: 0 completed with delta: 0.522 over: 2.000

[DEBUG] torsoPitchMj1 : set 1 : Reached zero position!

[DEBUG] |yarp.yri.Calibrator| torso-pitch-mj1-setup-calibrator finished calibration of device torso-pitch-mj1-setup-mc_remapper

[INFO] |yarp.yri.Robot| All actions for action level 10 of startup phase finished.

[INFO] |yarp.yri.Robot| startup phase finished.

[INFO] |yarp.yri.Robot| run phase starting...

[DEBUG] |yarp.yri| yarprobotinterface running happily

 
```
</details>


<details>
   <summary> YRI complains of Interfaces not implemented </summary>
   
```
ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
[ERROR] |yarp.device.controlBoard_nws_yarp| isJointBraked() failed
```
</details>


<details>
   <summary> MotorGui complains of Interfaces not implemented </summary>
   
```
[ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
[ERROR] Missing Implementation of checkMotionDone()
[WARNING] Missing Implementation of isJointBraked()
[ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
[ERROR] Missing Implementation of checkMotionDone()
[WARNING] Missing Implementation of isJointBraked()
[ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
[ERROR] Missing Implementation of checkMotionDone()
[WARNING] Missing Implementation of isJointBraked()
[ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
[ERROR] Missing Implementation of checkMotionDone()
[WARNING] Missing Implementation of isJointBraked()
```
</details>

cc @AntonioViscomi

AntonioViscomi commented:
> MotorGui complains of Interfaces not implemented
> ```
> [ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
> [ERROR] Missing Implementation of checkMotionDone()
> [WARNING] Missing Implementation of isJointBraked()
> [ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
> [ERROR] Missing Implementation of checkMotionDone()
> [WARNING] Missing Implementation of isJointBraked()
> [ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
> [ERROR] Missing Implementation of checkMotionDone()
> [WARNING] Missing Implementation of isJointBraked()
> [ERROR] |yarp.device.remote_controlboard| Unable to isJointBraked
> [ERROR] Missing Implementation of checkMotionDone()
> [WARNING] Missing Implementation of isJointBraked()
> ```

Hi @valegagge 
This problem might be fixed as reported in https://github.com/robotology/icub-main/issues/1047#issuecomment-3360788656. The fix has not yet been tested on a real setup

AntonioViscomi commented:
> The fix has not yet been tested on a real setup

I tested the fix on a setup and it works. https://github.com/AntonioViscomi/yarp/commit/52e5ac1cd188a07ec38ad952012dcd2b8a2b931e


---


 <details>
  <summary>wrong error in `calibrationDoneRaw`: in the end the joint is calibrated</summary>
  
> wrong error in `calibrationDoneRaw`: in the end the joint is calibrated
> ```
> DEBUG] |yarp.yri.Calibrator| torso-pitch-mj1-setup-calibrator starting calibration of device torso-pitch-mj1-setup-mc_remapper
> 
> [INFO] torsoPitchMj1 : starting calibration
> 
> [DEBUG] torsoPitchMj1 : Joints calibration order: 0
> 
> [WARNING] torsoPitchMj1  is calibrating only a subset of the robot part. Calibrating  1  over a total of  2
> 
> [DEBUG] torsoPitchMj1 : Calling calibrateJoint on joint  0 : type  12  with params:  9162 0 0 0 0
> 
> [DEBUG] from BOARD 10.0.1.1 (torso-pitch-mj1-setup) time=467s 555m 596u :  src LOCAL, adr 0,(code 0x04000001, par16 0x0000 par64 0x0000000000000000) -> DEBUG: tag01  CALIB 12 j 0: offset=41930 zero=-32768 
> 
> [DEBUG] torsoPitchMj1 : set 1 j 0 : Calibrating... enc values just before starting low-level calibration:  -2.80701
> 
> [ERROR] BOARD torso-pitch-mj1-setup (IP 10.0.1.1)  Unable to complete calibration: joint 0  is in  CALIBRATION  status in  virtual yarp::dev::ReturnValue yarp::dev::embObjMotionControl::calibrationDoneRaw(int)
> 
> [DEBUG] from BOARD 10.0.1.1 (torso-pitch-mj1-setup) time=468s 139m 65u :  src LOCAL, adr 0,(code 0x04000001, par16 0x0001 par64 0x0000000000000001) -> DEBUG: tag01  SET CONTROLMODE
> 
> [DEBUG] torsoPitchMj1 : calib joint  0 ended
> 
> [DEBUG] torsoPitchMj1 : set 1 : Calibration ended, going to zero!
> 
> [DEBUG] torsoPitchMj1 : Sending positionMove to joint 0  (desired pos:  0 desired speed:  30  )
> 
> [DEBUG] torsoPitchMj1: checkGoneToZeroThreshold: joint: 0 curr: -0.522 des: 0.000 -> delta: 0.522 threshold: 2.000 output: 319.000 mode: pos
> 
> [DEBUG] torsoPitchMj1: checkGoneToZeroThreshold: joint: 0 completed with delta: 0.522 over: 2.000
> 
> [DEBUG] torsoPitchMj1 : set 1 : Reached zero position!
> 
> [DEBUG] |yarp.yri.Calibrator| torso-pitch-mj1-setup-calibrator finished calibration of device torso-pitch-mj1-setup-mc_remapper
> 
> [INFO] |yarp.yri.Robot| All actions for action level 10 of startup phase finished.
> 
> [INFO] |yarp.yri.Robot| startup phase finished.
> 
> [INFO] |yarp.yri.Robot| run phase starting...
> 
> [DEBUG] |yarp.yri| yarprobotinterface running happily
> ```
 </details>

Fixed by means of https://github.com/AntonioViscomi/icub-main/commit/e11440f621887fd429cedfd3cd9ec732a28acf79


---

EDIT
The error below
<details>
 <summary>YRI crashes due to `yarp_telemetryDeviceDumper.so`</summary>
  
> YRI crashes due to `yarp_telemetryDeviceDumper.so`
> ```
> [ERROR] yarprobotinterface intercepted a segmentation fault caused by a faulty plugin:
> [ERROR] /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os7Vocab326encodeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x13) [0x71308eba5a33]
> Trace requested at /home/icub/projects/robotology-superbuild/src/YARP/src/commands/yarprobotinterface/Module.cpp:82 by code called from:
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_Z16yarp_print_traceP8_IO_FILEPKcj+0x3b) [0x71308eb3a6fb]
> yarprobotinterface(+0xb114) [0x6037a977f114]
> /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x71308e666520]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os7Vocab326encodeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x13) [0x71308eba5a33]
> /home/icub/projects/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN9robometry21TelemetryDeviceDumper22loadSettingsFromConfigERN4yarp2os10SearchableE+0x1c0a) [0x71308427f34a]
> /home/icub/projects/robotology-superbuild/build/install/lib/yarp/yarp_telemetryDeviceDumper.so(_ZN9robometry21TelemetryDeviceDumper4openERN4yarp2os10SearchableE+0x50) [0x713084280d90]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_dev.so.3(_ZN4yarp3dev10PolyDriver8coreOpenERNS_2os10SearchableE+0x691) [0x71308e5b54f1]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_dev.so.3(_ZN4yarp3dev10PolyDriver4openERNS_2os10SearchableE+0x50) [0x71308e5b6120]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface6Device4openEv+0x5c) [0x71308ecaa5ec]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface5Robot7Private11openDevicesEv+0x4c3) [0x71308ecb2af3]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_robotinterface.so.3(_ZN4yarp14robotinterface5Robot10enterPhaseENS0_11ActionPhaseE+0x637) [0x71308ecb4727]
> yarprobotinterface(+0xc127) [0x6037a9780127]
> /home/icub/projects/robotology-superbuild/build/install/lib/libYARP_os.so.3(_ZN4yarp2os8RFModule9runModuleERNS0_14ResourceFinderE+0xf7) [0x71308eb8f8d7]
> ```

 </details>

Might be related to the missing porting of robometry to YARP4. Here are some qdd fixes that makes robometry compatible with yarp4: https://github.com/AntonioViscomi/robometry/commit/7112d300cdd6c6e859af9258eb3ab8176565bed4
I haven't tested these fixes on the setup yet.

Also, it seems that we have installed yarp4 on top of yarp3, which creates symlinks that tricks robometry into loading yarp4 stuff when it is supposed to load yarp3.x stuff

```bash
icub@iiticublap244:~/projects/robotology-superbuild/build/install/lib$ ls -ls libYARP_os.so*
   0 lrwxrwxrwx 1 icub icub      15 Nov 15 00:38 libYARP_os.so -> libYARP_os.so.3
   0 lrwxrwxrwx 1 icub icub      19 Dec 10 10:51 libYARP_os.so.3 -> libYARP_os.so.4.0.0
2468 -rw-r--r-- 1 icub icub 2525424 Nov 15 00:31 libYARP_os.so.3.12.1
2340 -rw-r--r-- 1 icub icub 2396000 Dec 15 10:00 libYARP_os.so.4.0.0
 ```

cc @valegagge @MSECode @Nicogene

AntonioViscomi commented:
> Might be related to the missing porting of robometry to YARP4. Here are some qdd fixes that makes robometry compatible with yarp4: [AntonioViscomi/robometry@7112d30](https://github.com/AntonioViscomi/robometry/commit/7112d300cdd6c6e859af9258eb3ab8176565bed4) I haven't tested these fixes on the setup yet.

Fix done by applying the changes in the commit above and reinstalling robometry

AntonioViscomi commented:
During testing of the `velocity direct` control mode, the functionality appeared correct. and all other control modes operated as expected using the PID settings found in the configuration file.

I had to create a dedicated PID group for the velocity direct mode because the default configuration caused noticeable joint vibration. This exposed a critical behavior:  the joint began vibrating before the velocity direct control mode was actually enabled in the `ymotorgui` `view` menu.

I am unsure whether these values reset automatically following a fault, but verifying this behavior may be worthwhile. After applying the custom PID group, the vibration was resolved.

Before proceeding with checking the yarprobotinterface, I would address the `isJointBraked` error.

cc @valegagge

valegagge commented:
Well done @AntonioViscomi !
Tomorrow we can decide how to avoid to flooding the bash of yri and motorgui with ibrake message.

Note: yesterday we saw that the flooding behavior could be due to a bug in yarp. 
Maybe @randaz81 can help us!

valegagge committed:
YARP4 alignment and velocity direct mode

- Aligned all libraries to YARP4 API (embObjMotionControl, CanBusMotionControl,
  calibrators, dragonfly2, couplings)
- Implemented velocity direct control mode (parser, PID, setpoints, 2FOC)
- Removed framegrabbergui2

valegagge committed:
Dragonfly and code review fixes

- Fixed dragonfly2 returnValue handling
- Applied Gemini review suggestions to FirewireCameraDC1394
- Fixed EmbObjMotionControl issues (GetRefPosition, calibration error)
- Restored cmake files modified by velocity direct commit

valegagge committed:
embObjMotionControl refactoring: parser, defaults, measureConverter

- Refactored eomcParser with default values and paramValues namespace
- Converted string constants to std::string_view
- Modernized measureConvFactors: double* to std::vector<double>
- Implemented IdeviceDriveParams interface
- Removed unused macros and helper_pid functions
- Renamed private fields with _ prefix convention

valegagge committed:
YARP4 IControlMode enum migration: ControlModeEnum and SelectableControlModeEnum

valegagge commented on src/libraries/ctrlLib/include/iCub/ctrl/filters.h:
```diff
@@ -207,7 +207,7 @@ class RateLimiter : public IFilter
-    void getLimits(yarp::sig::Vector &rL, yarp::sig::Vector &rU);
+    void getPosLimits(yarp::sig::Vector &rL, yarp::sig::Vector &rU);
 
     /**
```
hi @gemini-code-assist,
I don't want to keep the recotrompability. These are all baking changes

valegagge commented on CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ cmake_minimum_required(VERSION 3.12...3.28)
 
 set(ROOT_PROJECT_NAME ICUB)
 project(${ROOT_PROJECT_NAME} LANGUAGES C CXX
-                             VERSION 2.11.1)
+                             VERSION 2.10.0)
 
```
Yes you are right. I need to increase it to 2.12.0!

valegagge commented on src/libraries/icubmod/couplingICubHandMk2/CouplingICubHandMk2.cpp:
```diff
@@ -306,8 +306,8 @@ bool CouplingICubHandMk2::convertFromActuatedAxesToPhysicalJointsAcc(const yarp:
     physJointsAcc[19] = actAxesAcc[8]/3;
     
-    return false;
+    return yarp::dev::ReturnValue::return_code::return_value_error_generic;
```
Yes, you are right!

valegagge replied:
Hi @Nicogene ,
This is the same question: 
<img width="1982" height="1097" alt="image" src="https://github.com/user-attachments/assets/c43d17ef-c99d-44b5-86d7-924e8db7142e" />

valegagge commented on src/libraries/icubmod/couplingICubHandMk2/CouplingICubHandMk2.cpp:
```diff
@@ -138,22 +138,22 @@ bool CouplingICubHandMk2::convertFromPhysicalJointsToActuatedAxesAcc(const yarp:
     // l/r_pinky
     actAxesAcc[8] = physJointsAcc[13] + physJointsAcc[14] + physJointsAcc[15];
-    return false;
+    return yarp::dev::ReturnValue::return_code::return_value_error_generic;
```
Yes, you are right!

valegagge replied:
Hi @Nicogene ,
as you can see above the copilot suggests me to change the return from false to a positive value. We have done it, but now I have a doubt: is it the expected behaviour? or you put false for a reason?

this is the comperasin between the devl branch and this:
<img width="1987" height="791" alt="image" src="https://github.com/user-attachments/assets/4c5fbac1-5586-4645-90e8-85644051ba8e" />


Thanks!

valegagge commented on src/tools/trajectoryPlayer/utils.cpp:
```diff
@@ -303,7 +303,7 @@ bool robotDriver::init()
     //set the initial reference speeds
     double* speeds = new double [n_joints];
     for (int i=0; i<n_joints; i++) speeds[i] = 20.0;
-    ipos_ll->setRefSpeeds(speeds);
+    ipos_ll->setTrajSpeeds(speeds);
 
```
Yes, you are right, but this PR doesn't aim to update all modules to moder c++. we keep this code as is.

valegagge commented on src/tools/CMakeLists.txt:
```diff
@@ -25,7 +25,6 @@ if(ICUB_USE_Qt5)
      message(STATUS "GSL not found")
    endif()
 
-   add_subdirectory(frameGrabberGui2)
    add_subdirectory(skinManagerGui)
    add_subdirectory(FirmwareUpdater)
 else()
```
I'll check it

valegagge replied:
I re-enable the module in [4bc9c24](https://github.com/robotology/icub-main/pull/1058/commits/4bc9c24fc6723d75543c4ceee8941c64485b0dc6)

valegagge committed:
Bumbed version to 2.12.0 and do two fies (copilot)

valegagge commented:
Hi @gemini-code-assist,
ca do a review and please focus your review on the changes in embObjMotioncontrol.cpp and its header file and on eomcParser.cpp and h. I'm interested in checking that their functionalities are not changed despite the code refactoring. Finally, tell me, if you have any suggestions regarding the new sw code structure af the above-mentioned files.

valegagge commented on src/modules/ctpService/main.cpp:
```diff
@@ -256,33 +256,34 @@ class scriptPosPort
         if (autochange_control_mode_enable)
         {
-            mode->setControlModes(disp.length(), joints.data(), modes.data());
+            std::vector<yarp::dev::SelectableControlModeEnum> setModes(disp.length(), yarp::dev::SelectableControlModeEnum::VOCAB_CM_POSITION);
+            mode->setControlModes(joints, setModes);
```
I don't want to update other modules except the device of yarp

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.h:
```diff
@@ -462,127 +364,134 @@ class yarp::dev::embObjMotionControl:   public DeviceDriver,
+    virtual ReturnValue setPidRaw(const PidControlTypeEnum& pidtype, int j, const Pid &pid) override;
+    virtual ReturnValue setPidsRaw(const PidControlTypeEnum& pidtype, const Pid *pids) override;
+    virtual ReturnValue setPidReferenceRaw(const PidControlTypeEnum& pidtype, int j, double ref) override;
+    virtual ReturnValue setPidReferencesRaw(const PidControlTypeEnum& pidtype, const double *refs) override; // Qui non so se refs is null o contine la memoria dimensionata correttamente
```
Yes, you are right, but the IPidControl yarp interface exposes these interfaces.

valegagge commented:
Hi @gemini-code-assist,
please review the code.

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.h:
```diff
@@ -227,6 +192,16 @@ using namespace iCub;
  */
+
+ /********************************************************************/
+ /********************************************************************/
+/*       IMPORTANT IMPLEMENTATION DETAILS */
+/* Each virtual method nimplemented in this device shoudl NOT check 
+if the provided joint is in the managed joint range or if the given vector ahs been allocated with the correct size. 
+This checks are performed by the higher-level interfaces.
+*/
```
ok

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControlConfigTypes.h:
```diff
@@ -0,0 +1,556 @@
+
+        yInfo("  fbk_PidUnits: %s", fbkUnitsToString(fbk_PidUnits).c_str());
+        yInfo("  out_PidUnits: %s", outUnitsToString(out_PidUnits).c_str());
+        yInfo("  out_type: %s", outTypeToString(out_type).c_str());
```
This TODO will be addressed in the future

valegagge commented on src/libraries/icubmod/dragonfly2/linux/FirewireCameraDC1394-DR2_2.cpp:
```diff
@@ -2087,78 +2104,79 @@ bool CFWCamera_DR2_2::setPowerDC1394(bool onoff)
     int fps=maxFPS(videoMode,colorCoding);
-    double bpp=bytesPerPixel(colorCoding);
-    double maxBandOcc=double(fps*xDim*yDim)*bpp;
+    double bytesPerPixelValue=bytesPerPixel(colorCoding); // And here used for computation. Should we rename this to e.g. "double bpPixel?
```
In this PR I don't address this type of improvement

valegagge committed:
copilot'suggestions fixes

valegagge commented:
Hi @gemini-code-assist,
please review the code.

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.cpp:
```diff
@@ -472,6 +407,9 @@ bool Parser::parseSelectedVelocityControl(yarp::os::Searchable &config) // OK
 {
+    //TODO: understand how to save the values of mixed control
+    return true;
+/*
```
Miked control mode fixed

valegagge committed:
preset ENABLE_icubmod_rawvaluespublisherremapper

valegagge committed:
Add frameGrabberGui2 and fix

I deleted the folder by mistake

API: qualify scoped enum types (cameraFeature_id_t, FeatureMode, BusType) with constexpr aliases for C++17 compatibility, migrate DC1394 getter calls from return-value to output-parameter style, and update IFrameGrabberControls calls from pointer to reference semantics

valegagge commented:
Hi @gemini-code-assist,
please review the changes

valegagge commented on CMakePresets.json:
```diff
@@ -45,7 +45,6 @@
         "ENABLE_icubmod_rawValuesPublisherClient": "ON",
         "ENABLE_icubmod_rawValuesPublisherServer": "ON",
-        "ENABLE_icubmod_rawValuesPublisherRemapper": "ON",
         "ENABLE_icubmod_fineCalibrationChecker": "ON"
```
done

valegagge committed:
minor fixes (copilot)

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.h:
```diff
@@ -268,116 +244,66 @@ class yarp::dev::embObjMotionControl:   public DeviceDriver,
 
 #ifdef NETWORK_PERFORMANCE_BENCHMARK 
-    Tools:Emb_RensponseTimingVerifier m_responseTimingVerifier;
+    Tools:Emb_RensponseTimingVerifier _responseTimingVerifier;
```
I don't care of this

valegagge commented:
Hi @gemini-code-assist,
please review the changes

valegagge commented on app/fineCalibrationChecker/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# CopyPolicy: Released under the terms of the GNU GPL v2.0.
+
+set(appname fineCalibrationChecker)
+file(GLOB 
+    conf ${CMAKE_CURRENT_SOURCE_DIR}/conf/*.ini
+    utl ${CMAKE_CURRENT_SOURCE_DIR}/*.csv
+)
```
I don't care of this

valegagge commented:
Hi @gemini-code-assist,
please review the changes

valegagge commented:
This PR is related to : https://github.com/robotology/icub-firmware-shared/pull/124

valegagge commented:
# 📓  Some additional notes regarding the embObjMotionControl refactoring 📓 

## 💡  Idea behind the refactoring
This PR includes a refactoring of embObjMotionControl mainly aimed at improving maintainability and making configuration handling more explicit (beyond the YARP4 interface changes).

- Default values were moved into a dedicated header (embObjMotionControlDefaults.h) to avoid scattered hard-coded fallbacks and keep defaults centralized.

- The data exchanged between eomcParser and the device was formalized in a shared types header (embObjMotionControlConfigTypes.h), clarifying the parser→device boundary.

- The code was modernized by removing C-like arrays in favor of safer/more idiomatic C++ containers.

- Configuration is now applied in two phases: the device reads configuration data from the parser and sends it to the embedded boards; after the initial setup it relies on the values stored in the network variables, and no longer uses (or expects updates from) the original configuration files.


## ✏️  Documentation regarding the default value.
[embObjMotionControlConfigurationDefaults.md](https://github.com/user-attachments/files/27125997/embObjMotionControlConfigurationDefaults.md)

I don't yet know where and if to add the `embObjMotionControlConfigurationDefaults.md` file to the user. Now we have the description of each parameter in the template folder in robots-configuration. I don't know if `embObjMotionControlConfigurationDefaults.md` helps the user or adds only confusion.

So for now, I decided to attach it here and report its content hereafter:

<details>

<summary> embObjMotionControlConfigurationDefaults.md </summary>
--------------------------------------------------------------


# embObjMotionControl Configuration Defaults (User Guide)

This document summarizes the default values used by embObjMotionControl when optional parameters are missing in configuration files.

Scope:
- Fallback values defined in the defaults singleton.
- Fallback behavior implemented by the parser.
- Practical notes for writing robust config files.

## 1. How defaults are applied

- If a parameter/group is optional and missing, embObjMotionControl uses internal fallback values.
- If a parameter/group is mandatory and missing, parsing fails.
- Some defaults are conditional (for example based on board type or encoder type).

## 2. Allowed string tokens (validation values)

These are accepted values for specific string parameters:

- `NONE`
- `low_lev_current`
- `direct`
- `minjerk`
- `torque`
- `metric_units`
- `machine_units`
- `dutycycle_percent`
- `PT100`
- `PT1000`
- `prismatic`
- `revolute`

## 3. Defaults by configuration group

## 3.1 FOC group

Optional parameters and fallbacks:

- `TemperatureSensorType`: `NONE`
- `Verbose`: `0` (disabled)
- `AutoCalibration`: `0` (disabled)
- `Kbemf`: `0.0`

Notes:
- `HasTempSensor` is deprecated. Use `TemperatureSensorType`.
- If `TemperatureSensorType` is omitted, firmware temperature checks are disabled.

## 3.2 TIMEOUTS group

If `TIMEOUTS` is missing, these defaults are applied to all joints:

- `velocity`: 100 ms
- `current`: 100 ms
- `pwm`: 100 ms
- `torque`: 100 ms
- `torque_measure`: 100 ms

If only one timeout entry is missing, only that entry falls back to default.

AMC FOC safeguard:
- For advanced FOC service boards, each timeout is clamped to a minimum of 300 ms.
- Effective minimums are:
  - `velocity`: 300 ms
  - `current`: 300 ms
  - `pwm`: 300 ms
  - `torque`: 300 ms
  - `torque_measure`: 300 ms

## 3.3 LIMITS group (temperature)

Optional parameters and fallbacks:

- `hardwareTemperatureLimits`: 0.0
- `warningTemperatureLimits`: 0.0

Validation rule:
- `warningTemperatureLimits` must be less than or equal to 85% of `hardwareTemperatureLimits`.

## 3.4 Torque PID optional terms

When torque control is enabled, these torque compensation terms are optional and default to 0.0 if omitted:

- `kbemf`
- `viscousPos`
- `viscousNeg`
- `coulombPos`
- `coulombNeg`
- `velocityThres`

Mandatory (no fallback):
- `ktau`
- `filterType`
- standard PID terms required by the selected control law.

## 3.5 Position-direct and velocity-direct PID blocks

If `CONTROLS.positionDirect` or `CONTROLS.velocityDirect` is set to `NONE`, the corresponding PID is disabled.

Fallback behavior:
- Position direct PID: disabled.
- Velocity direct PID: disabled.
- Stored PID values are treated as inactive for those control modes.

## 3.6 KALMAN_FILTER group

If the whole group is missing, Kalman filtering is disabled for all joints and these defaults are used:

- `kalmanFilterEnabled`: `false`
- `x0`: `[0.0, 0.0, 0.0]`
- `Q`: `[0.0, 0.0, 0.0]`
- `R`: `0.0`
- `P0`: `0.0`

If the group exists, all listed fields are expected.

## 3.7 OTHER_CONTROL_PARAMETERS.deadZone

If `OTHER_CONTROL_PARAMETERS` or `deadZone` is missing, deadzone is set by joint encoder type:

- AEA encoder: 0.0494
- AEA3 encoder: 0.0
- AKSIM2 encoder: 0.00068
- AMO encoder: 0.0055
- Other encoders: intended fallback is 0.00068

## 3.8 LUGRE group

If the `LUGRE` group is missing:

- `Km` defaults to `-1.0`.
- Other LuGre parameters are not explicitly assigned by this fallback path.

If the group is present, all required LuGre fields must be provided.

## 4. Additional internal defaults (runtime behavior)

These are not direct parser fallbacks for configuration fields, but they affect runtime behavior:

- `jointPosLimitsDelta = 2.0`
  - Returned position limits are reduced by this margin in limit reporting logic.
- `maxPositionMoveInterval = 0.080 s`
  - If positionMove commands arrive faster than this threshold, a performance warning is issued.

## 5. Quick checklist for configuration authors

- Always provide all mandatory groups first (`LIMITS`, `CONTROLS`, joint and motor definitions, PID mandatory parts).
- If you omit optional fields, verify that defaults are compatible with your hardware.
- For AMC advanced FOC boards, remember effective timeout minimum is 300 ms.
- Explicitly set `TemperatureSensorType` and temperature limits if thermal protection is required.
- If you use deadzone defaults, verify encoder type mapping is correct for each joint.


</details>
